### PR TITLE
Restore stateful crawling support

### PR DESCRIPTION
**Pull request by boolean5**
This PR restores support for stateful crawling by addressing the core issue that broke it, i.e., the fact that geckodriver would delete the browser profile when closing or crashing before we had a chance to archive it. It does so by using a [custom profile](https://firefox-source-docs.mozilla.org/testing/geckodriver/CrashReports.html) instead of creating the profile via the `FirefoxProfile` Selenium class. The custom profile is set via the arguments of the `Options` class. 

**Note:**
Geckodriver has a bug that makes it write the browser preferences we set, as well as its own default browser preferences, to a `user.js` file in the wrong profile directory when using a custom profile: https://github.com/mozilla/geckodriver/issues/1844. As a temporary workaround until this issue gets fixed, we create the `user.js` file ourselves. In order to do this, we keep a copy of geckodriver's default preferences in `configure_firefox.py`.

This PR also:
* Fixes a bug in `PatchedGeckoDriverService` class that caused geckodriver not to receive the `service_args` passed when starting the browser.
* Removes the [code that intercepts the location of the browser profile from the geckodriver logs](https://github.com/mozilla/OpenWPM/blob/cb95ecc05fb4618e4275faecbc62b84123ebf1b0/openwpm/deploy_browsers/selenium_firefox.py#L71L74), as we no longer need it.
* Updates `manual_test.py`

Closes https://github.com/mozilla/OpenWPM/issues/423, Closes https://github.com/mozilla/OpenWPM/issues/383, Closes https://github.com/mozilla/OpenWPM/issues/161, Closes https://github.com/mozilla/OpenWPM/issues/253, Closes https://github.com/mozilla/OpenWPM/issues/384, Closes https://github.com/mozilla/OpenWPM/issues/680

### CODE REVIEW

File: openwpm/browser_manager.py [hunk 1]
```diff
@@ -394,33 +401,32 @@ def shutdown_browser(self, during_init: bool, force: bool = False) -> None:
...
                 close_webdriver=False,
-                compress=True
+                compress=True,
+            )
+            dump_profile_command.execute(
+                webdriver=None,
+                browser_params=self.browser_params,
+                manager_params=self.manager_params,
+                extension_socket=None,
             )
```
Comment (boolean5) [inline openwpm/browser_manager.py:429]: As I understand this, we can't use `DumpProfileCommand` as part of a command sequence here and in `launch_browser_manager`. Our only choice is to call its `execute()` method directly. Passing `None` to  `webdriver` and `extension_socket` looks a bit weird though. I also had to change the type of `extension_socket` from `ClientSocket` to `Optional[ClientSocket]` in `DumpProfileCommand`'s `execute()` method for mypy to stop complaining.
Do you have a suggestion on how to do this in a better way?

**Comment by boolean5** [comment]
`manual_test.py` [still sets](https://github.com/mozilla/OpenWPM/blob/cb95ecc05fb4618e4275faecbc62b84123ebf1b0/test/manual_test.py#L117) the browser profile via the `FirefoxProfile` class. Should we change it to use a custom profile instead?
My impression is that since we're not interested in being able to archive the profile in the context of `manual_test.py`, there's no need to change it, but I'd like a confirmation on this.

**Comment by vringar** [comment]
Yes, we should use a custom profile Here.
Good job for remembering `manual_test.py`!

File: test/test_profile.py [hunk 2]
```diff
@@ -1,83 +1,81 @@
...
-            manager_params.data_directory, "browser_profile"
-        )
-        return manager_params, browser_params
+def test_saving(default_params, task_manager_creator):
+    manager_params, browser_params = default_params
+    manager_params.num_browsers = 1
+    browser_params[0].profile_archive_dir = (
+        manager_params.data_directory / "browser_profile"
+    )
+    manager, _ = task_manager_creator((manager_params, browser_params[:1]))
```
[the remark this answers is not in the document]
Comment (vringar) [reply test/test_profile.py:20]: I mostly did this to make sure our TaskManager tests weren't broken for more than one browser.
Maybe I should have set the default to 1.
I think running with 1 browser here is the right thing to do.

File: openwpm/browser_manager.py (hunk 1 above)
Comment (vringar) [reply openwpm/browser_manager.py:429]: Hmm, not really.
Maybe we can extract the actual implementation of the DumpProfileCommand.execute method into a free function and then call it from here and the Command.

Comment (englehardt) [reply openwpm/browser_manager.py:429]: > Maybe we can extract the actual implementation of the DumpProfileCommand.execute method into a free function and then call it from here and the Command.

This sounds like the right solution to me.

File: openwpm/deploy_browsers/configure_firefox.py [hunk 3]
```diff
@@ -1,7 +1,103 @@
 """ Set prefs and load extensions in Firefox """
 
+import json
+import re
+from pathlib import Path
+from typing import Any, Dict
+
+# TODO: Remove hardcoded geckodriver default preferences once
```
Comment (englehardt) [inline openwpm/deploy_browsers/configure_firefox.py:8]: We should add a step into our [release checklist](https://github.com/mozilla/OpenWPM/blob/master/docs/Release-Checklist.md) to check whether these prefs should be updated whenever a new version of geckodriver is released. We should also have an issue on file to remove the hack once the fix ships.

File: openwpm/deploy_browsers/deploy_firefox.py [hunk 4]
```diff
@@ -110,16 +117,30 @@ def deploy_firefox(
         # TODO restore detailed logging
         # fo.set_preference("extensions.@openwpm.sdk.console.logLevel", "all")
 
+    # Geckodriver currently places the user.js file in the wrong profile
+    # directory, so we have to create it manually here.
+    # TODO: Remove this workaround once
```
Comment (englehardt) [inline openwpm/deploy_browsers/deploy_firefox.py:122]: File a new issue for this removal and reference it here. (We don't want to track TODOs in the codebase).

File: openwpm/deploy_browsers/deploy_firefox.py [hunk 5]
```diff
@@ -110,16 +117,30 @@ def deploy_firefox(
...
+    # Load existing preferences from the profile's user.js file
+    prefs = configure_firefox.load_existing_prefs(browser_profile_path)
+    # Load default geckodriver preferences
+    prefs.update(configure_firefox.DEFAULT_GECKODRIVER_PREFS)
+    # Pick an available port for Marionette (https://stackoverflow.com/a/2838309)
+    s = socket.socket()
+    s.bind(("", 0))
+    marionette_port = s.getsockname()[1]
+    s.close()
+    prefs["marionette.port"] = marionette_port
```
[the remark this answers is not in the document]
Comment (englehardt) [reply openwpm/deploy_browsers/deploy_firefox.py:133]: > The way I've implemented picking a port and passing it to geckodriver has a race condition, as another process may get the port before Marionette does. I think this will rarely happen though. Is it something we can accept? Can we think of a better way to do this?

This seems acceptable. I agree with Stefan that a comment noting this is possible will be helpful.

What happens if this fails? I assume we'll automatically retry?

File: openwpm/deploy_browsers/deploy_firefox.py [hunk 6]
```diff
@@ -129,16 +150,21 @@ def deploy_firefox(
...
     # Launch the webdriver
     status_queue.put(("STATUS", "Launch Attempted", None))
     fb = FirefoxBinary(firefox_path=firefox_binary_path)
     driver = webdriver.Firefox(
-        firefox_profile=fp,
         firefox_binary=fb,
-        firefox_options=fo,
+        options=fo,
         log_path=interceptor.fifo,
+        # TODO: Remove when https://github.com/mozilla/geckodriver/issues/1844
```
Comment (englehardt) [inline openwpm/deploy_browsers/deploy_firefox.py:165]: Again, better to link to our own issue here.

File: test/test_crawl.py [hunk 7]
```diff
@@ -1,18 +1,17 @@
-# type:ignore
-# As this file is no longer maintained, mypy shouldn't check this
-import os
+"""Runs a short test crawl.
```
Comment (englehardt) [inline test/test_crawl.py:1]: Check out https://github.com/mozilla/OpenWPM/issues/197. I disabled this test a while back because I don't love the idea of connecting to real websites during the tests.

Is there a way to restrict this test to only run on CI and not when a user runs the test suite locally?

File: test/test_crawl.py [hunk 8]
```diff
@@ -42,117 +41,99 @@ def get_public_suffix(url):
...
+        ps = get_public_suffix(url)
+        hist_ps.add(ps)
+        statuses[ps] = command_status
+
+    # Grab urls from Firefox database
+    profile_ps = set()  # visited domains from firefox profile
+    rows = db_utils.query_db(ff_db, "SELECT url FROM moz_places")
+    for (host,) in rows:
+        try:
+            profile_ps.add(get_public_suffix(host))
```
Comment (englehardt) [inline test/test_crawl.py:94]: Is this actually adding the public suffix (e.g., `com`) or the public suffix + 1 (e.g., `example.com`)? If it's the former, it seems like we should update this to use public suffix + 1. To grab that we can directly use `du.get_ps_plus_1`.

File: test/test_profile.py [hunk 9]
```diff
@@ -1,83 +1,81 @@
...
-        manager.get("example.com")  # Cause a selenium crash
-
-        # The browser will fail to launch due to the proxy crashes
-        try:
-            manager.get("http://example.com")
-        except CommandExecutionError:
-            pass
-        manager.close()
-        assert isfile(join(browser_params[0].profile_archive_dir, "profile.tar.gz"))
+@pytest.mark.skip(reason="proxy no longer supported, need to update")
```
Comment (englehardt) [inline test/test_profile.py:53]: Just a reminder that we also want to figure out how to fix this.

**Review by englehardt — CHANGES_REQUESTED** [review]
This is great! Thank you Georgia! I left a bunch of comments, but nothing major.

**Comment by boolean5** [comment]
> Yes, we should use a custom profile Here.
> Good job for remembering `manual_test.py`!

`manual_test.py` happened to be broken for a different reason and also needed another update (check f5bacaed84ae624b0a9d602bd4f28770976b8164)

File: openwpm/deploy_browsers/configure_firefox.py (hunk 3 above)
Comment (boolean5) [reply openwpm/deploy_browsers/configure_firefox.py]: Done! Here's the issue: https://github.com/mozilla/OpenWPM/issues/867

File: test/test_crawl.py (hunk 7 above)
Comment (boolean5) [reply test/test_crawl.py:1]: Sure there is, I added this

```python
@pytest.mark.skipif(
    "CI" not in os.environ or os.environ["CI"] == "false",
    reason="Makes remote connections",
)
```

**Comment by boolean5** [comment]
I assume that in these three cases [[1]](https://github.com/mozilla/OpenWPM/blob/b6849c71f53185d1202d862ed34ef71d9c8cc860/crawler.py#L223) [[2]](https://github.com/mozilla/OpenWPM/blob/b6849c71f53185d1202d862ed34ef71d9c8cc860/test/test_callback.py#L20) [[3]](https://github.com/mozilla/OpenWPM/blob/b6849c71f53185d1202d862ed34ef71d9c8cc860/test/test_profile.py#L92) `reset=True` was used because stateful crawling was disabled and we can now let it take the default `False` value. Is there a reason that I'm missing to keep `reset=True` for any of them?

**Comment by englehardt** [comment]
> I assume that in these three cases [[1]](https://github.com/mozilla/OpenWPM/blob/b6849c71f53185d1202d862ed34ef71d9c8cc860/crawler.py#L223) [[2]](https://github.com/mozilla/OpenWPM/blob/b6849c71f53185d1202d862ed34ef71d9c8cc860/test/test_callback.py#L20) [[3]](https://github.com/mozilla/OpenWPM/blob/b6849c71f53185d1202d862ed34ef71d9c8cc860/test/test_profile.py#L92) `reset=True` was used because stateful crawling was disabled and we can now let it take the default `False` value. Is there a reason that I'm missing to keep `reset=True` for any of them?

For [1] we should keep reset=True. We want those crawls to be stateless (of course other users can change as they see fit). For [2] and [3] I'm not sure it matters, but I'll defer to @vringar just in case.

**Comment by vringar** [comment]
[2] and [3] can be removed as well.

File: test/test_crawl.py (hunk 8 above)
Comment (boolean5) [reply test/test_crawl.py]: Good catch! This was adding the publix suffix indeed so I changed it to use `du.get_ps_plus_1`.
Also, after looking at the test more closely, I noticed that one of the queries to the crawl database was not formed correctly (see 9e8298c).

File: openwpm/deploy_browsers/deploy_firefox.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (boolean5) [reply openwpm/deploy_browsers/deploy_firefox.py]: Yes, the browser manager is automatically restarted

**Comment by boolean5** [comment]
As we discussed, I checked and the original tar file is kept intact after extracting, so I removed the intermediate step of [copying](https://github.com/mozilla/OpenWPM/blob/ab01a2f6bdadcf75fe8c2878bc8d6610f5ed2a7a/openwpm/commands/profile_commands.py#L131L141) it from `load_profile()`.

File: test/test_profile.py (hunk 9 above)
Comment (boolean5) [reply test/test_profile.py:53]: Let's track this in https://github.com/mozilla/OpenWPM/issues/789, so that it doesn't delay this PR. I'll add a comment there.

**Comment by boolean5** [comment]
@vringar, @englehardt, thanks for reviewing! This is ready, I've covered all the requested changes. You can take another look if you'd like.

File: test/test_profile.py [hunk 10]
```diff
@@ -89,7 +87,7 @@ def test_seed_persistance(default_params, task_manager_creator):
 
     command_sequences = []
     for _ in range(2):
-        cs = CommandSequence(url="https://example.com", reset=True)
+        cs = CommandSequence(url="https://example.com")
```
Comment (vringar) [inline test/test_profile.py:90]: Just realizing we're making an outbound connection here.
Could you change this to use our local server or should I do that in a different PR?

Comment (boolean5) [reply test/test_profile.py:90]: Hm, this is also true for `test_saving` and `test_crash_profile`. I'll push a commit to fix that. Thanks for pointing it out!

File: openwpm/deploy_browsers/selenium_firefox.py [hunk 11]
```diff
@@ -129,7 +122,7 @@ def __init__(
         self.service_args = service_args or []
 
     def command_line_args(self):
-        return ["--port", "%d" % self.port]
+        return ["--port", "%d" % self.port] + self.service_args
 
     def send_remote_shutdown_command(self):
```
Comment (vringar) [inline openwpm/deploy_browsers/selenium_firefox.py:127]: Do you know why we pass here? What are we trying to supress/stop from happening?

**Review by vringar — APPROVED** [review]
Do we delete the profile folders on a normal shutdown? I don't think we need to but maybe we should still do it.

File: openwpm/deploy_browsers/selenium_firefox.py (hunk 11 above)
Comment (boolean5) [reply openwpm/deploy_browsers/selenium_firefox.py:127]: Do you mean why we pass in `send_remote_shutdown_command()`? class `PatchedGeckoDriverService` is a patched version of [Selenium's `Service` class](https://github.com/SeleniumHQ/selenium/blob/0849604f66a3e15e7ae650cfb0ac870328ce271f/py/selenium/webdriver/firefox/service.py) and we've copied this method from there.

Comment (vringar) [reply openwpm/deploy_browsers/selenium_firefox.py:127]: Ah okay, thank you. 
Is there a reason we shouldn't inherit from the Firefox service and overwrite `__init__` instead of having to keep up with the changes in Selenium?

**Comment by boolean5** [comment]
> Do we delete the profile folders on a normal shutdown? I don't think we need to but maybe we should still do it.

I think we do that [here](https://github.com/mozilla/OpenWPM/blob/49965db164e424f174dd749a87902ad95e10db63/openwpm/browser_manager.py#L419L421), in `shutdown_browser()`. Does that cover what you had in mind?

**Comment by vringar** [comment]
> > Do we delete the profile folders on a normal shutdown? I don't think we need to but maybe we should still do it.
> 
> I think we do that [here](https://github.com/mozilla/OpenWPM/blob/49965db164e424f174dd749a87902ad95e10db63/openwpm/browser_manager.py#L419L421), in `shutdown_browser()`. Does that cover what you had in mind?

Yeah. Sorry for missing this and thanks for pointing it out.

File: openwpm/deploy_browsers/selenium_firefox.py (hunk 11 above)
Comment (boolean5) [reply openwpm/deploy_browsers/selenium_firefox.py:127]: I was wrong. `PatchedGeckoDriverService` is a subclass of the `Service` class in [`selenium.webdriver.common.service`](https://github.com/SeleniumHQ/selenium/blob/0849604f66a3e15e7ae650cfb0ac870328ce271f/py/selenium/webdriver/common/service.py) and not [`selenium.webdriver.firefox.service`](https://github.com/SeleniumHQ/selenium/blob/0849604f66a3e15e7ae650cfb0ac870328ce271f/py/selenium/webdriver/firefox/service.py) as I initially thought. That explains why it has to define the`command_line_args()` and `send_remote_shutdown_command()` methods.

I guess we could still do what you are suggesting though. Do you think we should?

Comment (vringar) [reply openwpm/deploy_browsers/selenium_firefox.py:127]: Yes, because this way we only need to check what changed in the `__init__` Method of `selenium.webdriver.firefox.service` and not all changes to that class.

**Review by vringar — APPROVED** [review]

**Review by englehardt — APPROVED** [review]

**Comment by birdsarah** [comment]
WOOOOTTTT!!!!

**Comment by vringar** [comment]
Not officially released as of yet though. Gonna drop it with the next FF release.
